370-auth-ui/session.md

```diff
@@ -0,0 +1,171 @@
+      and vulnerable to CSRF.
+    - Setting to `Strict` is possible, but not recommended since it may cause
+      misbehavior to due missing session cookie in initial requests.
+- `Domain` attribute is not set by default. This can be configured by developer:
+    - If not set, the session cookie is only accessible by the auth domain.
+      Sharing session across services is not possible. For example,
+      `accounts.example.com` cannot share session with `forum.example.com`.
+    - To share sessions across services, this attribute should be set to a
+      common suffix of the service domains. For the previous example, a
+      suitable suffix would be `example.com`.
+- Developer can configured whether the cookie is a HTTP session cookie:
+    - If it is configured as a HTTP session cookie, the cookie would be deleted
+      from client side when browser session ends.
+    - If it is configured as a HTTP permanent cookie, the cookie would be
+      persisted until the maximum lifetime of session is reached.
+
+#### Configuration
+
+Sample configuration:
+```yaml
+session:
+    lifetime: 86400
+    idle_timeout_enabled: true
+    idle_timeout: 300
+    cookie_same_site: None
+    cookie_domain: example.com
+    cookie_expiration: false
+clients:
+    - auth_api_use_cookie: false
+```
+
+- **session**: Configuration for the session mechanism.
+    - **lifetime**: The maximum session lifetime in seconds,
+                    default to 2592000 (30 days).
+    - **idle_timeout_enabled**: Whether session idle timeout is enabled,
+                                default to false.
+    - **idle_timeout**: The session idle timeout period in seconds,
+                        default to 300 (5 minutes).
+    - **cookie_same_site**: The `SameSite` attribute of cookie,
+                            default to `Lax`, and other possible values are
+                            `None` and `Strict`.
+    - **cookie_domain**: The `Domain` attribute of cookie, optional.
+    - **cookie_expiration**: Whether the session cookie would expire
+                             (i.e. HTTP permanent cookie), default to false.
+- **clients**: Configuration of OIDC clients. Details are in other specs.
+    - **auth_api_use_cookie**: Whether the Auth API would set session cookie,
+                               instead of returning OIDC access/refresh tokens,
+                               default to false.
+
+### Session Management
+
+#### List
+
+Skygear Auth provides an API to list all valid sessions and offline grants of
+the user. The list includes the session attributes.
+
+### Session Resolution
+
+Skygear Auth provides an API to resolve user info from cookie / OIDC access
+token.
+
+For cookies, a session can be directly resolved from the session cookie.
+
+For OIDC access token, its parent grant would be resolved:
+- if it is a access grant: the associated session would be resolved.
+- if it is a offline grant: the offline grant itself would be resolved.
```

Comment:
Can we clarify access grant vs offline grant?

Reply:
Here is an updated diagram on relationship between grants and session:

[![](https://mermaid.ink/img/eyJjb2RlIjoiY2xhc3NEaWFncmFtXG4gICAgQXBwIFwiMVwiIC0tPiBcIk1cIiBDbGllbnRcbiAgICBBcHAgXCIxXCIgLS0-IFwiTVwiIFVzZXJcbiAgICBDbGllbnQgXCIxXCIgLS0-IFwiTVwiIENsaWVudENyZWRlbnRpYWxHcmFudCA6IENsaWVudCBDcmVkZW50aWFsc1xuICAgIFVzZXIgXCIxXCIgLS0-IFwiTVwiIElkUFNlc3Npb25cbiAgICBDbGllbnQgXCIxXCIgLS0-IFwiTVwiIEF1dGhvcml6YXRpb25cbiAgICBVc2VyIFwiMVwiIC0tPiBcIk1cIiBBdXRob3JpemF0aW9uXG4gICAgQXV0aG9yaXphdGlvbiBcIjFcIiAtLT4gXCJNXCIgQ29kZUdyYW50IDogQXV0aG9yaXphdGlvbiBDb2RlXG4gICAgSWRQU2Vzc2lvbiBcIjFcIiAtLT4gXCJNXCIgQ29kZUdyYW50XG4gICAgQ29kZUdyYW50IFwiMVwiIC0tPiBcIjFcIiBBY2Nlc3NHcmFudCA6IHdlYiBhcHBzXG4gICAgQ29kZUdyYW50IFwiMVwiIC0tPiBcIjFcIiBPZmZsaW5lR3JhbnQgOiBvZmZsaW5lX2FjY2VzcyBzY29wZSAoaS5lLiBuYXRpdmUgYXBwcylcbiAgICBVc2VyIFwiMVwiIC0tPiBcIk1cIiBPZmZsaW5lR3JhbnQ6IEF1dGggQVBJXG4gICAgT2ZmbGluZUdyYW50IFwiMVwiIC0tPiBcIk1cIiBBY2Nlc3NHcmFudCA6IHJlZnJlc2gvaXNzdWUgYWNjZXNzIHRva2VuXG4gICAgXG4gICAgY2xhc3MgVXNlciB7XG4gICAgICAgIGlkOiBzdHJpbmdcbiAgICB9XG4gICAgY2xhc3MgQXV0aG9yaXphdGlvbiB7XG4gICAgICAgIGlkOiBzdHJpbmdcbiAgICAgICAgY2xpZW50SUQ6IHN0cmluZ1xuICAgICAgICB1c2VySUQ6IHN0cmluZ1xuICAgICAgICBzY29wZXM6IHN0cmluZ1tdXG4gICAgfVxuICAgIGNsYXNzIENsaWVudCB7XG4gICAgICAgIGNsaWVudElEOiBzdHJpbmdcbiAgICAgICAgY2xpZW50U2VjcmV0OiBzdHJpbmdcbiAgICB9XG4gICAgY2xhc3MgSWRQU2Vzc2lvbiB7XG4gICAgICAgIGlkOiBzdHJpbmdcbiAgICAgICAgdXNlcklEOiBzdHJpbmdcbiAgICAgICAgaWRlbnRpdHlJRDogc3RyaW5nXG4gICAgfVxuICAgIGNsYXNzIENsaWVudENyZWRlbnRpYWxHcmFudCB7XG4gICAgICAgIGNsaWVudElEOiBzdHJpbmdcbiAgICAgICAgY2xpZW50U2VjcmV0OiBzdHJpbmdcbiAgICAgICAgZXhwaXJlQXQ6IERhdGVcbiAgICAgICAgc2NvcGVzOiBzdHJpbmdbXVxuICAgICAgICBhY2Nlc3NUb2tlbjogc3RyaW5nXG4gICAgfVxuICAgIGNsYXNzIENvZGVHcmFudCB7XG4gICAgICAgIGF1dGh6SUQ6IHN0cmluZ1xuICAgICAgICBzZXNzaW9uSUQ6IHN0cmluZ1xuICAgICAgICBleHBpcmVBdDogRGF0ZVxuICAgICAgICBzY29wZXM6IHN0cmluZ1tdXG4gICAgICAgIGNvZGU6IHN0cmluZ1xuICAgIH1cbiAgICBjbGFzcyBBY2Nlc3NHcmFudCB7XG4gICAgICAgIGF1dGh6SUQ6IHN0cmluZ1xuICAgICAgICBzZXNzaW9uSUQ_OiBzdHJpbmdcbiAgICAgICAgZXhwaXJlQXQ6IERhdGVcbiAgICAgICAgc2NvcGVzOiBzdHJpbmdbXVxuICAgICAgICBhY2Nlc3NUb2tlbjogc3RyaW5nXG4gICAgfVxuICAgIGNsYXNzIE9mZmxpbmVHcmFudCB7XG4gICAgICAgIGF1dGh6SUQ6IHN0cmluZ1xuICAgICAgICBleHBpcmVBdDogRGF0ZVxuICAgICAgICBzY29wZXM6IHN0cmluZ1tdXG4gICAgICAgIHJlZnJlc2hUb2tlbjogc3RyaW5nXG4gICAgfVxuXG4iLCJtZXJtYWlkIjp7InRoZW1lIjoiZGVmYXVsdCJ9LCJ1cGRhdGVFZGl0b3IiOmZhbHNlfQ)](https://mermaid-js.github.io/mermaid-live-editor/#/edit/eyJjb2RlIjoiY2xhc3NEaWFncmFtXG4gICAgQXBwIFwiMVwiIC0tPiBcIk1cIiBDbGllbnRcbiAgICBBcHAgXCIxXCIgLS0-IFwiTVwiIFVzZXJcbiAgICBDbGllbnQgXCIxXCIgLS0-IFwiTVwiIENsaWVudENyZWRlbnRpYWxHcmFudCA6IENsaWVudCBDcmVkZW50aWFsc1xuICAgIFVzZXIgXCIxXCIgLS0-IFwiTVwiIElkUFNlc3Npb25cbiAgICBDbGllbnQgXCIxXCIgLS0-IFwiTVwiIEF1dGhvcml6YXRpb25cbiAgICBVc2VyIFwiMVwiIC0tPiBcIk1cIiBBdXRob3JpemF0aW9uXG4gICAgQXV0aG9yaXphdGlvbiBcIjFcIiAtLT4gXCJNXCIgQ29kZUdyYW50IDogQXV0aG9yaXphdGlvbiBDb2RlXG4gICAgSWRQU2Vzc2lvbiBcIjFcIiAtLT4gXCJNXCIgQ29kZUdyYW50XG4gICAgQ29kZUdyYW50IFwiMVwiIC0tPiBcIjFcIiBBY2Nlc3NHcmFudCA6IHdlYiBhcHBzXG4gICAgQ29kZUdyYW50IFwiMVwiIC0tPiBcIjFcIiBPZmZsaW5lR3JhbnQgOiBvZmZsaW5lX2FjY2VzcyBzY29wZSAoaS5lLiBuYXRpdmUgYXBwcylcbiAgICBVc2VyIFwiMVwiIC0tPiBcIk1cIiBPZmZsaW5lR3JhbnQ6IEF1dGggQVBJXG4gICAgT2ZmbGluZUdyYW50IFwiMVwiIC0tPiBcIk1cIiBBY2Nlc3NHcmFudCA6IHJlZnJlc2gvaXNzdWUgYWNjZXNzIHRva2VuXG4gICAgXG4gICAgY2xhc3MgVXNlciB7XG4gICAgICAgIGlkOiBzdHJpbmdcbiAgICB9XG4gICAgY2xhc3MgQXV0aG9yaXphdGlvbiB7XG4gICAgICAgIGlkOiBzdHJpbmdcbiAgICAgICAgY2xpZW50SUQ6IHN0cmluZ1xuICAgICAgICB1c2VySUQ6IHN0cmluZ1xuICAgICAgICBzY29wZXM6IHN0cmluZ1tdXG4gICAgfVxuICAgIGNsYXNzIENsaWVudCB7XG4gICAgICAgIGNsaWVudElEOiBzdHJpbmdcbiAgICAgICAgY2xpZW50U2VjcmV0OiBzdHJpbmdcbiAgICB9XG4gICAgY2xhc3MgSWRQU2Vzc2lvbiB7XG4gICAgICAgIGlkOiBzdHJpbmdcbiAgICAgICAgdXNlcklEOiBzdHJpbmdcbiAgICAgICAgaWRlbnRpdHlJRDogc3RyaW5nXG4gICAgfVxuICAgIGNsYXNzIENsaWVudENyZWRlbnRpYWxHcmFudCB7XG4gICAgICAgIGNsaWVudElEOiBzdHJpbmdcbiAgICAgICAgY2xpZW50U2VjcmV0OiBzdHJpbmdcbiAgICAgICAgZXhwaXJlQXQ6IERhdGVcbiAgICAgICAgc2NvcGVzOiBzdHJpbmdbXVxuICAgICAgICBhY2Nlc3NUb2tlbjogc3RyaW5nXG4gICAgfVxuICAgIGNsYXNzIENvZGVHcmFudCB7XG4gICAgICAgIGF1dGh6SUQ6IHN0cmluZ1xuICAgICAgICBzZXNzaW9uSUQ6IHN0cmluZ1xuICAgICAgICBleHBpcmVBdDogRGF0ZVxuICAgICAgICBzY29wZXM6IHN0cmluZ1tdXG4gICAgICAgIGNvZGU6IHN0cmluZ1xuICAgIH1cbiAgICBjbGFzcyBBY2Nlc3NHcmFudCB7XG4gICAgICAgIGF1dGh6SUQ6IHN0cmluZ1xuICAgICAgICBzZXNzaW9uSUQ_OiBzdHJpbmdcbiAgICAgICAgZXhwaXJlQXQ6IERhdGVcbiAgICAgICAgc2NvcGVzOiBzdHJpbmdbXVxuICAgICAgICBhY2Nlc3NUb2tlbjogc3RyaW5nXG4gICAgfVxuICAgIGNsYXNzIE9mZmxpbmVHcmFudCB7XG4gICAgICAgIGF1dGh6SUQ6IHN0cmluZ1xuICAgICAgICBleHBpcmVBdDogRGF0ZVxuICAgICAgICBzY29wZXM6IHN0cmluZ1tdXG4gICAgICAgIHJlZnJlc2hUb2tlbjogc3RyaW5nXG4gICAgfVxuXG4iLCJtZXJtYWlkIjp7InRoZW1lIjoiZGVmYXVsdCJ9LCJ1cGRhdGVFZGl0b3IiOmZhbHNlfQ)

Simply speaking, refresh token is only issued in two situations:
- Auth API is configured to return OIDC access & refresh token
- offline_access scope is included in Authorization Request, i.e. in native apps

In these cases, an offline grant would be created to be a container of the refresh token and necessary session attributes (e.g. IP, user agent).